x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceTests.java

```diff
@@ -479,6 +480,79 @@ public void testActionsForSystemUserIsAuthorized() {
         verifyNoMoreInteractions(auditTrail);
     }
 
+    public void testActionsForUserMatchingSystemUserRoleNameDenied() {
+        final Authentication authentication = createAuthentication(new User(SystemUser.NAME, SystemUser.ROLE_NAME));
+        final IndexRequest request = mock(IndexRequest.class);
+        final String requestId = AuditUtil.getOrGenerateRequestId(threadContext);
+
+        RoleDescriptor role = new RoleDescriptor(SystemUser.ROLE_NAME, new String[] {}, null, null, null, null, null, null);
+        roleMap.put(SystemUser.ROLE_NAME, role);
```

Comment:
I think this is not the right test for the changes. The change is to (1) allow user-defined roles to use internal names (CRUD operations) and also (2) a role name like `_system` should be resolved correctly to the user-define role, not the internal one. 

I believe the test here is more for (2). But the critical piece of "resolving role names to roleDescriptors" is mocked out with `roleMap` and `buildRole(...)`. So what it really exercises is the orchestration logic and also RBACEngine with given roleDescriptor. The better place to test "resolving role names to roleDescriptors" is `CompositeRolesStoreTests`

Reply:
The system user is special, in that there is additional logic in `AuthorizationService` that deals with it, before we even get to the role store:

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L255-L258

and

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L615-L625

I intended this and the second test to cover that logic, and the tests I added (or extended) in `CompositeRolesStoreTests` to handle the other resolving role name scenarios.

I agree that this test is less focused on role resolution and more on internal users and would have been better placed in my previous PR. Since it does give us additional coverage of a path that is closely related to role resolution I think it still makes sense to keep. WYT?

Reply:
> I agree that this test is less focused on role resolution and more on internal users and would have been better placed in my previous PR. Since it does give us additional coverage of a path that is closely related to role resolution I think it still makes sense to keep. WYT?

OK. It's indeed more in context for #86398. I have no problem for you to keep it here. My main point is to make sure there is a test for CompositeRolesStore that checks:
(1) A role name like `_xpack_security` will _not_ be resolved to the internal role definition
(2) An internal user such as `XpackSecurityUser` can still get its role resolved correctly regardless whether there is an user-defined role called `_xpack_security`.

Some of those probably already exist. But it would be great if you could verify.

Reply:
Makes sense! I've added added explicit tests for (1). 

For (2), `testDoesNotUseRolesStoreInternalUsers` verifies that there were no interactions with the stores, and `testGetRoleDescriptorsListForInternalUsers` does not even mock them; this should implicitly cover the assertion that resolution works regardless whether there is an user-defined role with a matching name.